pipeline_dp/dp_engine.py

```diff
@@ -57,6 +57,46 @@ def _add_report_stages(self, stages_description):
         for stage_description in stages_description:
             self._add_report_stage(stage_description)
 
+    def calculate_private_contribution_bounds(
+        self,
+        col,
+        params: pipeline_dp.CalculatePrivateContributionBoundsParams,
+        data_extractors: pipeline_dp.DataExtractors,
+        partitions: Any,
```

Comment:
So, actually, we will need to get the number of partitions. But if it is PCollection, it will be not trivial, but doable. I guess we can do something similar to what histograms computation does. We calculate the number and get PCollection<Integer> and then we calculate the histogram and get PCollection<Histogram>. And then I guess in python it will be possible to flatten it even though they have different types (if not we can create type HistogramOrInt :) ?): backend.flatten(histogram, number). And then we will have PCollection<Object> and we will call backend.to_list(pCollection) and then we will be able to apply map function, something like lambda el: el[0] is histogram, el[1] is number of partitions.
Does it make sense to you?

Reply:
Yeah, I agree that backend.flatten(histogram, number) approach would work and it seems to be a good approach. And no worries about the type, it's Python :).

Since otherwise we would need to ask user to compute the number of partitions, So we have a tradeoff between user convenience and complexity of the implementation. Since complexity is not high, here I'm for computing the count by ourselves.

Reply:
Nice, then I will send you implementation next week!